### ISSUE TITLE
유니티 가상환경에서의 경로 탐색 ML agent

### ISSUE THREAD

**Issue by choBottle** [HBNU-SWUNIV/come-capstone26-physicalAI#297]
## 개요
Unity ML-Agents를 활용하여 GO1 로봇의 디지털 트윈 환경에서
목표 지점까지 장애물을 회피하며 자율 주행하는 강화학습 에이전트 구현

---

## 시스템 구조
현실 GO1 로봇
- ↓ 카메라 + YOLO 장애물 탐지
- Unity 디지털 트윈 환경
- ↓ ML-Agents PPO 강화학습
- 목표 지점까지 장애물 회피 경로 탐색

---

### 1. GO1Agent.cs (강화학습 에이전트)
- PPO 알고리즘 기반 강화학습
- 관측값 9개
  - 목표 방향 벡터 (3)
  - 목표 거리 정규화 (1)
  - 에이전트 전방 벡터 (3)
  - 목표 정렬도 dot (1)
  - 회전 방향 힌트 cross (1)
- 행동: 전진/후진, 좌/우 회전 (Continuous 2개)
- 보상 설계
  - 목표 접근 시 +보상
  - 목표 도달 시 +10
  - 매 스텝 -0.001 (빠른 도달 유도)
  - 벽에 막힘(stuck) 시 -패널티

### 2. TrainingManager.cs (학습 환경 관리)
- GO1 에이전트 5개 자동 생성
- 각 에이전트별 Physics Layer 분리 → 상호 충돌 무시
- NavMesh 기반 랜덤 목표 지점 자동 생성
- 에이전트 목표 도달 시 새 목표 자동 할당

### 3. MapObstacleAutoSetup.cs (맵 장애물 자동 설정)
- 맵 내 오브젝트 이름 키워드 기반 자동 처리
  - wall, desk, chair, sofa, leg 등
- BoxCollider 자동 추가
- NavMeshObstacle 자동 추가 (carving)
- Obstacle 태그 자동 설정

### 4. Ray Perception Sensor 3D
- go1 프리팹에 추가
- Rays Per Direction: 5
- Max Ray Degrees: 120
- Ray Length: 5
- Detect Tags: Obstacle
- 장애물 감지 → 회피 경로 학습 가능

---

### Physics Layer
| Layer | 이름 |
|-------|------|
| 7 | Agent0 |
| 8 | Agent1 |
| 9 | Agent2 |
| 10 | Agent3 |
| 11 | Agent4 |

Agent 간 Layer Collision Matrix 전부 해제

### ML-Agents yaml 설정
```yaml
behaviors:
  GO1Agent:
    trainer_type: ppo
    hyperparameters:
      batch_size: 1024
      buffer_size: 10240
      learning_rate: 3.0e-4
      beta: 5.0e-3
      epsilon: 0.2
      lambd: 0.95
      num_epoch: 3
      learning_rate_schedule: linear
    network_settings:
      normalize: true
      hidden_units: 256
      num_layers: 3
    reward_signals:
      extrinsic:
        gamma: 0.99
        strength: 1.0
      curiosity:
        gamma: 0.99
        strength: 0.02
        encoding_size: 64
        learning_rate: 3.0e-4
    max_steps: 5000000
    time_horizon: 1024
    summary_freq: 10000
    keep_checkpoints: 5
```

### ML-Agents yaml 설정
```yaml
behaviors:
  GO1Agent:
    trainer_type: ppo          # 학습 알고리즘 (PPO: 안정적인 강화학습 방법)
    hyperparameters:
      batch_size: 1024         # 한 번에 학습할 데이터 수 (클수록 안정적)
      buffer_size: 10240       # 학습 전 수집할 경험 데이터 총량
      learning_rate: 3.0e-4   # 학습 속도 (너무 크면 불안정, 너무 작으면 느림)
      beta: 5.0e-3             # 탐색 장려 정도 (높을수록 다양한 행동 시도)
      epsilon: 0.2             # 정책 업데이트 허용 범위 (안정성 조절)
      lambd: 0.95              # 장기/단기 보상 균형 (1에 가까울수록 장기 중시)
      num_epoch: 3             # 수집한 데이터로 몇 번 반복 학습할지
      learning_rate_schedule: linear  # 학습률을 학습이 진행될수록 점점 줄임
    network_settings:
      normalize: true          # 입력 관측값 정규화 (학습 안정성 향상)
      hidden_units: 256        # 신경망 은닉층 뉴런 수 (클수록 복잡한 패턴 학습)
      num_layers: 3            # 신경망 깊이 (깊을수록 복잡한 행동 학습 가능)
    reward_signals:
      extrinsic:               # 코드에서 직접 설정한 보상
        gamma: 0.99            # 미래 보상 할인율 (1에 가까울수록 먼 미래 보상 중시)
        strength: 1.0          # 외부 보상 가중치
      curiosity:               # 호기심 기반 내부 보상 (탐색 촉진)
        gamma: 0.99            # 호기심 보상 할인율
        strength: 0.02         # 호기심 보상 가중치 (너무 크면 목표 무시하고 탐색만 함)
        encoding_size: 64      # 호기심 모델 인코딩 크기
        learning_rate: 3.0e-4  # 호기심 모델 학습률
    max_steps: 5000000         # 전체 학습 최대 스텝 수 (500만 스텝)
    time_horizon: 1024         # 한 에피소드에서 수집할 최대 스텝 수
    summary_freq: 10000        # 10000 스텝마다 학습 결과 출력
    keep_checkpoints: 5        # 저장할 체크포인트 모델 수
```
---

## 학습 실행 방법

1. 터미널에서 학습 시작
```bash
mlagents-learn config/GO1Agent.yaml --run-id=go1_run_05 --torch-device cuda
```

2. Unity에서 Play 버튼 클릭

3. 학습 모니터링
```bash
nvidia-smi -l 1
```

---

## 학습 결과 (go1_run_01 기준 / 빈 맵)

| Step | Mean Reward |
|------|-------------|
| 10K | -3.39 |
| 50K | -0.15 |
| 100K | +12.02 |
| 200K | +12.72 |
| 310K | +12.75 (수렴) |

약 100K step에서 수렴 확인

---

## 향후 계획
- [ ] Ray Perception Sensor 기반 장애물 회피 학습 완료
- [ ] 현실 GO1 YOLO 장애물 탐지 연동
- [ ] ObstacleDetection.cs로 실시간 장애물 반영
- [ ] 학습 완료 모델 실제 로봇 적용

**Comment by choBottle** [comment]
## 진행 내용

### 1. M키 기반 ML-Agent 경로 생성 흐름 구현
- Unity에서 M키 입력 시 가상 GO1이 ML-Agent 정책에 따라 목표 지점까지 이동하도록 구성하였다.
- 가상 GO1이 이동하는 동안 실제 GO1은 움직이지 않고, Unity 내부에서만 경로를 생성하도록 구성하였다.
- 목표 도착 후 가상 GO1이 지나간 경로를 기록하여 C++ 제어 코드로 JSON 형태로 전달하는 흐름을 구현하였다.

### 2. 현재 pose 기반 연속 경로 생성 구조 적용
- 첫 번째 경로 이동이 끝난 뒤, 다음 M키 입력 시 현재 실제 GO1의 위치와 각도를 새로운 시작 기준으로 사용하도록 수정하였다.
- 기존처럼 항상 초기 방향으로 복귀하는 방식이 아니라, 예를 들어 실제 GO1이 130도 방향으로 도착하면 다음 경로도 해당 위치와 각도를 기준으로 시작할 수 있도록 구성하였다.
- 이를 통해 여러 목표 지점을 순차적으로 이동하는 테스트 구조를 만들었다.

### 3. 가상 GO1 경로 생성 속도 개선
- M키로 가상 GO1이 경로를 생성할 때만 이동 속도를 빠르게 적용하도록 수정하였다.
- 현실 GO1의 이동 속도는 C++ 제어 코드에서 유지하고, 가상 경로 탐색 단계만 빠르게 수행하도록 분리하였다.
- `virtualPlanSpeedMultiplier`, `virtualPlanRotateMultiplier` 값을 추가하여 Inspector에서 조절 가능하도록 하였다.

### 4. 경로 기록 및 단순화 로직 조정
- 가상 GO1이 이동한 raw path를 기록하고, C++로 전달할 waypoint path를 별도로 생성하였다.
- `pathRecordInterval`, `minMoveThreshold`, `simplifyTolerance`, `minSendPointSpacing`, `maxSendPointCount`, `cornerKeepAngleDeg` 값을 통해 경로 점 개수를 조절할 수 있도록 하였다.
- 좁은 통로에서 경로가 지나치게 단순화되어 실제 이동 경로가 달라지는 문제를 확인하고, waypoint 보존 설정을 조정하였다.

### 5. 실제 전송 waypoint 시각화
- Unity에서 가상 GO1이 실제로 지나간 raw path와 C++로 전달되는 waypoint path를 구분하여 표시할 수 있도록 하였다.
- raw path와 전송 path의 차이를 확인하여, 경로 생성 단계 문제인지 C++ 추종 단계 문제인지 구분할 수 있도록 하였다.

### 6. 실제 GO1 이동 경로 시각화
- C++ state를 받아 Unity에서 driveByState로 움직이는 가상 GO1의 위치를 샘플링하여 실제 GO1의 이동 경로를 점 단위로 표시하도록 수정하였다.
- 실제 GO1이 계획된 경로 대비 어느 정도 벗어나는지 Unity 상에서 확인할 수 있게 되었다.
- 테스트 결과 특정 경로에서 Unity location 기준 약 0.3 정도의 위치 오차가 발생하는 것을 확인하였다.

### 7. 경로 막힘 판단 로직 검토
- M키 입력 전 NavMesh 기준으로 목표까지 경로가 존재하는지 확인하는 방식을 검토하였다.
- `NavMesh.CalculatePath()`의 `PathInvalid`, `PathPartial` 상태를 이용해 목표까지 도달 가능 여부를 판단할 수 있도록 하였다.
- 장애물이 NavMesh에 반영되지 않는 경우를 대비해 CapsuleCast 기반 물리 장애물 검사 방식도 검토하였다.
- 이동 중 실제 이동량이 거의 없을 경우 stuck 상태로 판단하는 로직도 포함하였다.

## 현재 확인된 문제

- Unity에서 생성한 waypoint path와 실제 GO1의 이동 경로 사이에 오차가 존재한다.
- 특히 실제 GO1이 C++ 내부 추정 위치보다 덜 이동하는 현상이 있으며, Unity location 기준 약 0.3 정도의 오차가 확인되었다.
- 이는 C++에서 명령 속도 기반 dead reckoning으로 `world_x`, `world_z`를 추정하기 때문에 발생하는 것으로 판단된다.

## 다음 작업

- C++의 위치 적분식에 실제 이동 보정 계수 적용 검토
- 실제 GO1 이동거리와 C++ 추정 이동거리 비교
- `odom_linear_scale` 또는 유사 보정값 적용
- waypoint 추종 방식을 단순 waypoint 추종에서 선분 기반 추종 또는 pure pursuit 방식으로 개선 검토
